NIO support for TMP_DIR

magicDGS opened the pull request:
Extracted from #3998

pshapiro4broad commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/markduplicates/MarkDuplicatesGATK.java:
```diff
@@ -358,7 +360,7 @@ private void generateDuplicateIndexes() {
                 (double) (Integer.MAX_VALUE - 5));
         logger.info("Will retain up to " + maxInMemory + " duplicate indices before spilling to disk.");
-        this.duplicateIndexes = new SortingLongCollection(maxInMemory, TMP_DIR.toArray(new File[TMP_DIR.size()]));
+        this.duplicateIndexes = new SortingLongCollection(maxInMemory, (Path[]) TMP_DIR.stream().map(IOUtils::getPath).toArray());
```
Instead of a cast, you can provide an argument to `toArray()`, e.g. `TMP_DIR.stream().map(IOUtils::getPath).toArray(Path[]::new)`

magicDGS replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineProgram.java:
```diff
@@ -60,8 +61,8 @@
-    @Argument(common=true, optional=true)
-    public List<File> TMP_DIR = new ArrayList<>();
+    @Argument(common=true, optional=true, doc = "List of temp directories to use.")
+    public List<String> TMP_DIR = new ArrayList<>();
```
Can you kebabify this argument as part of this PR (`--temp-dir`)? Looks like we missed this one.

droazen replied:
Actually, we just did a usage check on this argument, and it's ONLY used in `MarkDuplicates`. Given this, @lbergelson and I think that we should just remove this argument completely. Would you be ok with removing it @magicDGS, or are you currently depending on it?

magicDGS replied:
I depend on it, but I can move it to the tools that require it.

lbergelson replied:
@droazen Do we think a lot of users use this instead `-Djava.io.tmpdir`?  Maybe it's worth keeping a shortcut too?  Sorry to reconsider.

magicDGS replied:
I was reconsidering it too (that's why I didn't add a commit), because maybe there are other codepaths using teh java property, which is set here, instead of directly using the parameter. Next commit will not have this fix.

magicDGS replied:
But the name was changed on `StandardArgumentDefinitions.TMP_DIR_NAME` after rebase.

droazen commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineProgram.java:
```diff
@@ -151,13 +156,33 @@ public Object instanceMainPostParseArgs() {
+                logger.warn("Temp directory {}: unable to set permissions due to {}", p, e.getMessage());
+            }
+            // TODO - this should be p.toAbsolutePath().toUri().toString() to allow other FileSystems to be used (see above)
+            System.setProperty("java.io.tmpdir", p.toAbsolutePath().toString()); // in loop so that last one takes effect
```
@lbergelson suggests a special case here for file URIs to strip the `file://` prefix.

magicDGS replied:
Done in the next commit.

magicDGS commented:
Rebased to resolve conflicts and addressing comments.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineProgram.java:
```diff
@@ -151,13 +153,32 @@ public Object instanceMainPostParseArgs() {
+                // only set the permissions if they change
+                if (permissions.addAll(Arrays.asList(
+                        PosixFilePermission.OWNER_READ, PosixFilePermission.GROUP_READ, PosixFilePermission.OTHERS_READ,
+                        PosixFilePermission.OWNER_WRITE, PosixFilePermission.GROUP_WRITE, PosixFilePermission.OTHERS_WRITE))) {
```
By the way, this is related with https://github.com/broadinstitute/gatk/issues/4513 - should probably resolve here.

droazen commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineProgram.java:
```diff
@@ -60,8 +61,8 @@
-    @Argument(fullName = StandardArgumentDefinitions.TMP_DIR_NAME, common=true, optional=true)
-    public List<File> TMP_DIR = new ArrayList<>();
+    @Argument(fullName = StandardArgumentDefinitions.TMP_DIR_NAME, common=true, optional=true, doc = "List of temp directories to use.")
+    public List<String> TMP_DIR = new ArrayList<>();
```
If we're going to keep this argument in GATK, can you make the type into a `String` instead of `List<String>`, since it's not doing anything intelligent right now with the List?

magicDGS replied:
If I modify that, it changes the behavior of `MarkDuplicatesGATK` and `EstimateLibraryComplexityGATK` (I guess that they are setting the temporary directory properly)

droazen replied:
We can add a separate argument to those tools if they really need to take a List of temporary directories. But having the GATK-wide argument take a List and then ignore all but the last element is confusing.

magicDGS replied:
True. Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineProgram.java:
```diff
@@ -151,13 +153,32 @@ public Object instanceMainPostParseArgs() {
+                if (permissions.addAll(Arrays.asList(
+                        PosixFilePermission.OWNER_READ, PosixFilePermission.GROUP_READ, PosixFilePermission.OTHERS_READ,
+                        PosixFilePermission.OWNER_WRITE, PosixFilePermission.GROUP_WRITE, PosixFilePermission.OTHERS_WRITE))) {
+                    Files.setPosixFilePermissions(p, permissions);
```
Can you get rid of this code to change the permissions on the temp dir, and throw a `UserException` instead if the temp dir is not writable/readable?

magicDGS replied:
Done, although in that case is a change of behavior - if the temp dir is not used at all, this can fail for every tool...

droazen replied:
I think we always use the temp dir in GATK -- if for nothing else, then to extract and load the GKL library for compression/decompression.

magicDGS replied:
Is the GKL library _**always**_  extracted, even if not used at all? I don't think that it is (or should be) like that...

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/io/IOUtils.java:
```diff
@@ -557,6 +557,17 @@ public static Path getPath(String uriString) {
+     * @param path path to get the absolute name.
+     * @return a String with the absolute name.
+     */
+    public static String getAbsolutePathName(final Path path) {
```
The name of this method doesn't really tell you what it's doing -- can you rename to `getAbsolutePathWithoutFileProtocol()`?

magicDGS replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/io/IOUtils.java:
```diff
@@ -557,6 +557,17 @@ public static Path getPath(String uriString) {
+     * the file:// prefix.
+     *
+     * @param path path to get the absolute name.
+     * @return a String with the absolute name.
```
"a String with the absolute name, and the file:// protocol removed, if it was present"

magicDGS replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineProgram.java:
```diff
@@ -151,13 +153,32 @@ public Object instanceMainPostParseArgs() {
+                try {
+                    Files.createDirectories(p);
+                } catch (IOException e) {
+                    // intentionally ignoring
```
Don't create the temp dir if it doesn't exist -- just throw a `UserException` in this case

magicDGS replied:
I personally like specifying a non-existent directory for temporary files that is created by the program instead of explicitly using `mkdirs` - anyway, I can change in the next commit.

magicDGS replied:
Removed in the next commit.

droazen commented on src/test/java/org/broadinstitute/hellbender/utils/io/IOUtilsUnitTest.java:
```diff
@@ -196,6 +196,20 @@ private void innerTestGetPath(String s) throws IOException {
+        Assert.assertEquals(IOUtils.getAbsolutePathName(IOUtils.getPath(uriString)), expected);
+    }
+
     @Test
```
Can you add a unit test to prove that the temp dir works over NIO?

magicDGS replied:
Which kind of test do you need there? Retrieving the property and writing to it? Or running tools using it with NIO (e.g., `MarkDuplicatesGATK`)? Let me know for next round of comments...

droazen replied:
How about a test that:

1. Declares a dummy `CommandLineProgram` that executes `createTempFile()` and then asserts that it exists in its `doWork()` method

2. Runs the above command line program with `--tmp-dir` set to a jimfs URI. You can create a jimfs like this:

```
        try (FileSystem jimfs = Jimfs.newFileSystem(Configuration.unix())) {
            final Path tmpDirPath = jimfs.getPath("tmp");
```

3. You should also include a test case that runs the dummy `CommandLineProgram` with `--tmp-dir` set to a `file://` URI, since that is a different code path.

You could put the test in a new `CommandLineProgramIntegrationTest` class

magicDGS replied:
Which `createTempFile` should be use in the test? There are several implementations (htsjdk, gatk, `Files`, `File`), and all of them have their own peculiarities. The ones that work with a `File` object won't work with this change. Thus, I am not sure how to proceed here - for the tools using the tmpDir this should work because the codebase uses the htsjdk `Path` implementation, but otherwise I am not sure how to get this done well...

On the other hand,  there is a test class already for CLP (`CommandLineProgramUnitTest`) - should I still create the `*IntegrationTest` class or just use that one?

magicDGS commented:
Back to you, still without tests for tmp dir working with NIO due to lack of information: should it be at the tool level (e.g., running `MarkDuplicatesGATK` with NIO tmp dir - my main use case) or just checking the property value that is correctly set?

magicDGS commented:
@droazen - please, clarify how the tests should be done. I'm going to be on vacation next week, so feel free to add them to this branch if you would like to speed up the process (taking into account that #3998 depends on this now). Thanks!

magicDGS commented:
Friendly ping here @droazen

droazen commented on src/test/java/org/broadinstitute/hellbender/utils/io/IOUtilsUnitTest.java:
```diff
@@ -196,6 +196,20 @@ private void innerTestGetPath(String s) throws IOException {
+        return new Object[][] {
+                {"/local/example.txt", "/local/example.txt"},
+                {"/local/file://example.txt", "/local/file:/example.txt"},
+                {"file:///local/example.txt", "/local/example.txt"}
```
Add a test case with a non-file URI (such as `gcs://`)

magicDGS replied:
I guess that you mean adding the `gs://` example, that is the prefix in GCS. Otherwise it fails because `IOUtils.getPath` does not recognize it as a non-file URI.

Done and test passing.

droazen commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineProgram.java:
```diff
@@ -141,8 +138,8 @@ public final Object runTool(){
-        if (this.TMP_DIR == null) this.TMP_DIR = new ArrayList<>();
-        if (this.TMP_DIR.isEmpty()) TMP_DIR.add(IOUtil.getDefaultTmpDir());
+        // TODO - this should use the HTSJDK IOUtil.getDefaultTmpDirPath, which is somehow broken in the current HTSJDK version
+        if (TMP_DIR == null || TMP_DIR.isEmpty()) TMP_DIR = IOUtils.getAbsolutePathWithoutFileProtocol(IOUtils.getPath(System.getProperty("java.io.tmpdir")));
```
Curly braces around if statement body

magicDGS replied:
Leftover from previous implementation. Changed as it is clearer.

droazen commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineProgram.java:
```diff
@@ -60,8 +57,8 @@
-    @Argument(fullName = StandardArgumentDefinitions.TMP_DIR_NAME, common=true, optional=true)
-    public List<File> TMP_DIR = new ArrayList<>();
+    @Argument(fullName = StandardArgumentDefinitions.TMP_DIR_NAME, common=true, optional=true, doc = "Temp directory to use.")
+    public String TMP_DIR;
```
Can you rename this variable to `tmpDir` (but keep the argument name itself as-is)?

magicDGS replied:
Done

droazen commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineProgram.java:
```diff
@@ -151,13 +148,20 @@ public Object instanceMainPostParseArgs() {
+            // TODO: it may be that the program does not need a tmp dir
+            // TODO: if it fails, the problem can be discovered downstream
+            // TODO: should log a warning instead?
+            throw new UserException.BadInput(String.format("--%s (%s) should exists and have read/write access",
```
Can you use `UserException.BadTmpDir` instead (here and below)?

droazen replied:
Also, "should exists" should be "should exist"

lbergelson replied:
@magicDGS You probably want to rebase before making this change since I just made some changes to BadTmpDir

magicDGS replied:
Using `BadTmpDir` as currently implemented is not really pointing out where the error is: if the property was not set, and fails on checking the access it will point to a different tmp directory. I added a new constructor for the exception with a `Path` to provide the one in use.

Done.

magicDGS commented:
@droazen - Still some questions about integration tests (on the comment with your suggestion, but here too):

* I am not sure if the test that you are proposing will work with all the implementations of `createTempFile`: depends on how it is handle, as a `File` or as a `Path`
* I think that this depends a lot on the parts of the codebase that we are looking at, so maybe before accepting this a pass should be done for the usages and how the `java.io.tmpdir` is used.

Waiting for your feedback before doing something that does not make sense...

droazen commented:
@magicDGS Could you add overloads of `IOUtils.createTempFile()` and `BaseTest.createTempFile()` that take a Path, and use them just in the tests you write for this PR? Once tests are in place, we can gradually migrate usages to the Path overloads in future PRs.

magicDGS commented:
Sorry for the delay here, @droazen - I had some technical problems with my hardware and did not have time to work on this. Back to you with implemented tests!

droazen commented:
@magicDGS Could you rebase this branch one last time? Once it's been rebased, I'll give it a final look and (hopefully) hit merge. Thanks!

magicDGS commented:
Rebased @droazen - back to you!

droazen commented on src/test/java/org/broadinstitute/hellbender/cmdline/CommandLineProgramIntegrationTest.java:
```diff
@@ -0,0 +1,98 @@
+/*
+ * The MIT License (MIT)
```
Remove the license from this file. GATK4 has a repository-wide BSD 3-clause license.

magicDGS replied:
Removed (sorry, my IDE attach sometimes old LICENSE config...)

droazen commented on src/test/java/org/broadinstitute/hellbender/cmdline/CommandLineProgramIntegrationTest.java:
```diff
@@ -0,0 +1,98 @@
+/**
+ * @author Daniel Gomez-Sanchez (magicDGS)
+ */
+public class CommandLineProgramIntegrationTest extends GATKBaseTest {
```
Integration test classes should extend `CommandLineProgramTest` rather than `GATKBaseTest` directly.

magicDGS replied:
Done

droazen commented on src/test/java/org/broadinstitute/hellbender/cmdline/CommandLineProgramIntegrationTest.java:
```diff
@@ -0,0 +1,98 @@
+ */
+public class CommandLineProgramIntegrationTest extends GATKBaseTest {
+
+    private static FileSystem jimfs;
```
Create the `Jimfs` filesystem within your test case rather than at the class level. Static variables in test code tend to cause problems down the line.

magicDGS replied:
I found a solution by creating a new method in `GATKBaseTest` to close `Jimfs` for a `Path`. I refactored to remove this static field.

magicDGS replied:
For whatever reason, that method fails - it looks like GATK's `IOUtil.getPath` does not recognize `Jimfs` if the instance is created in the test method. Again, no action here.

droazen commented on src/test/java/org/broadinstitute/hellbender/cmdline/CommandLineProgramIntegrationTest.java:
```diff
@@ -0,0 +1,98 @@
+    public Object[][] tmpDirs() throws Exception {
+        return new Object[][]{
+                {createTempDir("local").toPath()},
+                {Files.createDirectory(jimfs.getPath("tmp"))}
```
There should be 3 test cases here rather than 2 in order to cover all the branches in your code: a local directory without a `file://` scheme specified, a local directory with `file://` specified, and a Jimfs URI.

magicDGS replied:
Actually, the one with the `file://` prefix is already tested - in the test, the argument is converted with `Path.toUri().toString()`, which adds the `file://`. I refactored the method to has the `String` argument directly, and added a test case for a local file without prefix.

droazen commented on src/test/java/org/broadinstitute/hellbender/cmdline/CommandLineProgramIntegrationTest.java:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    @Test(dataProvider = "tmpDirs", singleThreaded = true)
+    public void testCreateTempPath(final Path tmpDir) {
```
This test should be named something like `testTmpDirArgument`

magicDGS replied:
Done

droazen commented on src/test/java/org/broadinstitute/hellbender/utils/io/IOUtilsUnitTest.java:
```diff
@@ -166,6 +166,21 @@ private void innerTestGetPath(String s) throws IOException {
+    public Object[][] absoluteNames() {
+        return new Object[][] {
+                {"/local/example.txt", "/local/example.txt"},
+                {"/local/file://example.txt", "/local/file:/example.txt"},
```
This is a strange one... can you explain why `"/local/file://example.txt"` becomes `"/local/file:/example.txt"`?

magicDGS replied:
Due to normalization of the path (removal of redundant `/`) - I added a comment here to clarify for the future.

droazen commented on src/test/java/org/broadinstitute/hellbender/cmdline/CommandLineProgramIntegrationTest.java:
```diff
@@ -0,0 +1,98 @@
+
+        final Path p = (Path) new TestCreateTempPathClp().instanceMain(new String[]{
+                "--" + StandardArgumentDefinitions.TMP_DIR_NAME, tmpDir.toUri().toString()});
+        Assert.assertTrue(Files.exists(p));
```
Also assert that the `java.io.tmpdir` property has been changed.

magicDGS replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/io/IOUtils.java:
```diff
@@ -437,6 +439,40 @@ public static File createTempFileInDirectory(final String name, String extension
+     *
+     * @return A file in the temporary directory starting with name, ending with extension, which will be deleted after the program exits.
+     */
+    public static Path createTempPath(String name, String extension) {
```
Would be good to have a standalone unit test for this method in `IOUtilsUnitTest` (since the test in `CommandLineProgramIntegrationTest` is testing the `--tmp-dir` argument rather than this method directly.

magicDGS replied:
Done.

droazen commented on src/test/java/org/broadinstitute/hellbender/utils/io/IOUtilsUnitTest.java:
```diff
@@ -166,6 +166,21 @@ private void innerTestGetPath(String s) throws IOException {
 
+    @DataProvider
+    public Object[][] absoluteNames() {
+        return new Object[][] {
```
Can you also include a test case with a non-absolute path?

magicDGS replied:
Done

magicDGS commented on src/test/java/org/broadinstitute/hellbender/cmdline/CommandLineProgramIntegrationTest.java:
```diff
@@ -0,0 +1,98 @@
+        Assert.assertTrue(Files.exists(p));
+
+        // get back to the previous tmp dir
+        System.setProperty("java.io.tmpdir", previousTmpDir);
```
I also put this in a `finally` clause to be sure that after the test it always resets the previous tmp directory

droazen commented on src/test/java/org/broadinstitute/hellbender/cmdline/CommandLineProgramIntegrationTest.java:
```diff
@@ -0,0 +1,83 @@
+        try {
+            final Path p = (Path) new TestCreateTempPathClp().instanceMain(new String[] {
+                    "--" + StandardArgumentDefinitions.TMP_DIR_NAME, tmpDirArg});
+            Assert.assertTrue(Files.exists(p));
```
Should also assert here that the parent directory of p is equal to the `tmpDirArg`

magicDGS replied:
Done

droazen commented on src/test/java/org/broadinstitute/hellbender/cmdline/CommandLineProgramIntegrationTest.java:
```diff
@@ -0,0 +1,83 @@
+            final Path p = (Path) new TestCreateTempPathClp().instanceMain(new String[] {
+                    "--" + StandardArgumentDefinitions.TMP_DIR_NAME, tmpDirArg});
+            Assert.assertTrue(Files.exists(p));
+            Assert.assertNotEquals(System.getProperty("java.io.tmpdir"), previousTmpDir);
```
Ideally you would check here that `java.io.tmpdir` has been set to `tmpDirArg`, not just that it's different from `previousTmpDir`

magicDGS replied:
Done

droazen commented on src/test/java/org/broadinstitute/hellbender/GATKBaseTest.java:
```diff
@@ -119,6 +123,16 @@ public void initGenomeLocParser() throws FileNotFoundException {
+    /**
+     * Closes the {@link java.nio.file.FileSystem} for a path if it is an instance from {@link Jimfs}
+     */
+    public static void closeFileSystemIfJimnfs(final Path path) throws IOException {
```
I'm not sure this method should be up in `GATKBaseTest`, since you reported that you weren't able to use it in `CommandLineProgramIntegrationTest`. You should use the same method to create the Jimfs instances across your different test classes. At this point I'd be satisfied if you just deleted this method, and used the `@BeforeClass`/`@AfterClass` approach to set up the Jimfs in `IOUtilsUnitTest`.

magicDGS replied:
Deleted method.

droazen commented on src/test/java/org/broadinstitute/hellbender/cmdline/CommandLineProgramIntegrationTest.java:
```diff
@@ -0,0 +1,83 @@
+ */
+public class CommandLineProgramIntegrationTest extends CommandLineProgramTest {
+
+    private static FileSystem jimfs;
```
Can you make this non-static?

magicDGS replied:
Done

magicDGS commented:
@droazen - I also rebased locally to check the conflicting files: the differences are a new constant (`DEFAULT_TOOLKIT_SHORT_NAME`) and documentation for `TMP_DIR` (which is also added in this PR, but as it changed the `List` to `String`, it is a bit different). Do you want that I push the rebased branch, or do you prefer to wait until the end of the review?

magicDGS commented:
@droazen - friendly ping here!

droazen commented:
@magicDGS It looks like all of my comments here have been addressed, but the branch is in conflict. If you do a last rebase I can (finally) hit merge!

magicDGS commented:
Rebased - waiting for test passing and hopefully you can hit merge @droazen!

magicDGS commented:
Thanks for the review @droazen!
